UnsatisfiedLinkError of swt-win32-4965r8.dll on Windows Server 2008 R2

Hello,

Issue description: 

Recently we switched to the  latest Eclipse SDK release 
https://download.eclipse.org/releases/2024-06  version build [202405031000](https://download.eclipse.org/releases/2024-06/202405031000) 
and bumped with product crash on specific OS's:  Windows 8 / Windows Server 2008 
due to **UnsatisfiedLinkError of swt-win32-4965r8.dll**

PS:
Let us please, know if you reduce support of OS like Windows 8 / Windows 2008 RC  

The original issue in DBeaver product:
https://github.com/dbeaver/dbeaver/issues/34034

Hello, 
we switched from 
<eclipse-version>2024-06/202404121000/</eclipse-version>
to
<eclipse-version>2024-06/**202405031000**/</eclipse-version>

The entire support of Windows 7, Windows 8, Windows Server 2008 R2, MacOS Catalina 10.15.7
is broken due to   
```
Caused by: java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
	D:\bin\dbeaver\home\swt-win32-4965r8.dll: Belirtilen yordam bulunamadı
	no swt-win32 in java.library.path:
```
References issues 
https://github.com/dbeaver/dbeaver/issues/34244
https://github.com/dbeaver/dbeaver/issues/34197
https://github.com/dbeaver/dbeaver/issues/34034
https://github.com/dbeaver/dbeaver/issues/30700

Could you please confirm that  `Windows 7, Windows 8, Windows Server 2008 R2, MacOS Catalina 10.15.7`
out of support?

The issue happens on the most recent update of Windows Server 2016 which is still supported by Microsoft:

```
Caused by: java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
	C:\Program Files\DBeaver\swt-win32-4966r5.dll: The specified procedure could not be found
	no swt-win32 in java.library.path: 
```

--- a/bundles/org.eclipse.swt/Eclipse SWT PI/win32/library/os.c
+++ b/bundles/org.eclipse.swt/Eclipse SWT PI/win32/library/os.c
@@ -3276,7 +3276,15 @@ JNIEXPORT jint JNICALL OS_NATIVE(GetSystemMetricsForDpi)
 {
 	jint rc = 0;
 	OS_NATIVE_ENTER(env, that, GetSystemMetricsForDpi_FUNC);
+/*
 	rc = (jint)GetSystemMetricsForDpi(arg0, arg1);
+*/
+	{
+		OS_LOAD_FUNCTION(fp, GetSystemMetricsForDpi)
+		if (fp) {
+			rc = (jint)((jint (CALLING_CONVENTION*)(jint, jint))fp)(arg0, arg1);
+		}
+	}
 	OS_NATIVE_EXIT(env, that, GetSystemMetricsForDpi_FUNC);
 	return rc;
 }
@@ -3374,7 +3382,15 @@ JNIEXPORT jlong JNICALL OS_NATIVE(GetThreadDpiAwarenessContext)
 {
 	jlong rc = 0;
 	OS_NATIVE_ENTER(env, that, GetThreadDpiAwarenessContext_FUNC);
+/*
 	rc = (jlong)GetThreadDpiAwarenessContext();
+*/
+	{
+		OS_LOAD_FUNCTION(fp, GetThreadDpiAwarenessContext)
+		if (fp) {
+			rc = (jlong)((jlong (CALLING_CONVENTION*)())fp)();
+		}
+	}
 	OS_NATIVE_EXIT(env, that, GetThreadDpiAwarenessContext_FUNC);
 	return rc;
 }
@@ -6616,7 +6632,15 @@ JNIEXPORT jlong JNICALL OS_NATIVE(OpenThemeDataForDpi)
 	jlong rc = 0;
 	OS_NATIVE_ENTER(env, that, OpenThemeDataForDpi_FUNC);
 	if (arg1) if ((lparg1 = (*env)->GetCharArrayElements(env, arg1, NULL)) == NULL) goto fail;
+/*
 	rc = (jlong)OpenThemeDataForDpi((HWND)arg0, (LPCWSTR)lparg1, arg2);
+*/
+	{
+		OS_LOAD_FUNCTION(fp, OpenThemeDataForDpi)
+		if (fp) {
+			rc = (jlong)((jlong (CALLING_CONVENTION*)(HWND, LPCWSTR, jint))fp)((HWND)arg0, (LPCWSTR)lparg1, arg2);
+		}
+	}
 fail:
 	if (arg1 && lparg1) (*env)->ReleaseCharArrayElements(env, arg1, lparg1, JNI_ABORT);
 	OS_NATIVE_EXIT(env, that, OpenThemeDataForDpi_FUNC);
@@ -8865,7 +8889,15 @@ JNIEXPORT jlong JNICALL OS_NATIVE(SetThreadDpiAwarenessContext)
 {
 	jlong rc = 0;
 	OS_NATIVE_ENTER(env, that, SetThreadDpiAwarenessContext_FUNC);
+/*
 	rc = (jlong)SetThreadDpiAwarenessContext((DPI_AWARENESS_CONTEXT)arg0);
+*/
+	{
+		OS_LOAD_FUNCTION(fp, SetThreadDpiAwarenessContext)
+		if (fp) {
+			rc = (jlong)((jlong (CALLING_CONVENTION*)(DPI_AWARENESS_CONTEXT))fp)((DPI_AWARENESS_CONTEXT)arg0);
+		}
+	}
 	OS_NATIVE_EXIT(env, that, SetThreadDpiAwarenessContext_FUNC);
 	return rc;
 }
@@ -9215,7 +9247,15 @@ JNIEXPORT jboolean JNICALL OS_NATIVE(SystemParametersInfoForDpi)
 	jboolean rc = 0;
 	OS_NATIVE_ENTER(env, that, SystemParametersInfoForDpi_FUNC);
 	if (arg2) if ((lparg2 = getNONCLIENTMETRICSFields(env, arg2, &_arg2)) == NULL) goto fail;
+/*
 	rc = (jboolean)SystemParametersInfoForDpi(arg0, arg1, lparg2, arg3, arg4);
+*/
+	{
+		OS_LOAD_FUNCTION(fp, SystemParametersInfoForDpi)
+		if (fp) {
+			rc = (jboolean)((jboolean (CALLING_CONVENTION*)(jint, jint, NONCLIENTMETRICS *, jint, jint))fp)(arg0, arg1, lparg2, arg3, arg4);
+		}
+	}
 fail:
 	if (arg2 && lparg2) setNONCLIENTMETRICSFields(env, arg2, lparg2);
 	OS_NATIVE_EXIT(env, that, SystemParametersInfoForDpi_FUNC);

--- a/bundles/org.eclipse.swt/Eclipse SWT PI/win32/library/os_custom.h
+++ b/bundles/org.eclipse.swt/Eclipse SWT PI/win32/library/os_custom.h
@@ -20,3 +20,8 @@
 /* Libraries for dynamic loaded functions */
 #define GetDpiForMonitor_LIB "shcore.dll"
 #define RtlGetVersion_LIB "ntdll.dll"
+#define OpenThemeDataForDpi_LIB "uxtheme.dll"
+#define GetSystemMetricsForDpi_LIB "user32.dll"
+#define GetThreadDpiAwarenessContext_LIB "user32.dll"
+#define SetThreadDpiAwarenessContext_LIB "user32.dll"
+#define SystemParametersInfoForDpi_LIB "user32.dll"

--- a/bundles/org.eclipse.swt/Eclipse SWT PI/win32/org/eclipse/swt/internal/win32/OS.java
+++ b/bundles/org.eclipse.swt/Eclipse SWT PI/win32/org/eclipse/swt/internal/win32/OS.java
@@ -2990,6 +2990,7 @@ public static final native long GetSysColorBrush (int nIndex);
 /** @param hWnd cast=(HWND) */
 public static final native long GetSystemMenu (long hWnd, boolean bRevert);
 public static final native int GetSystemMetrics (int nIndex);
+/** @method flags=dynamic */
 public static final native int GetSystemMetricsForDpi (int nIndex, int dpi);
 /** @param hDC cast=(HDC) */
 public static final native int GetTextColor (long hDC);
@@ -3784,6 +3785,7 @@ public static final native boolean OpenClipboard (long hWndNewOwner);
  */
 public static final native long OpenThemeData (long hwnd, char[] pszClassList);
 /**
+ * @method flags=dynamic
  * @param hwnd cast=(HWND)
  * @param pszClassList cast=(LPCWSTR),flags=no_out
  */
@@ -4354,8 +4356,12 @@ public static final native int SetPixel (long hdc, int X, int Y, int crColor);
 /** @param hdc cast=(HDC) */
 public static final native int SetPolyFillMode (long hdc, int iPolyFillMode);
 public static final native boolean SetProcessDPIAware ();
-/** @param dpiContext cast=(DPI_AWARENESS_CONTEXT) */
+/**
+ * @method flags=dynamic
+ * @param dpiContext cast=(DPI_AWARENESS_CONTEXT)
+ */
 public static final native long SetThreadDpiAwarenessContext (long dpiContext);
+/** @method flags=dynamic */
 public static final native long GetThreadDpiAwarenessContext ();
 /** @method flags=no_gen */
 public static final native int SetPreferredAppMode(int mode);
@@ -4462,6 +4468,7 @@ public static final native boolean SystemParametersInfo (int uiAction, int uiPar
 public static final native boolean SystemParametersInfo (int uiAction, int uiParam, RECT pvParam, int fWinIni);
 public static final native boolean SystemParametersInfo (int uiAction, int uiParam, NONCLIENTMETRICS pvParam, int fWinIni);
 public static final native boolean SystemParametersInfo (int uiAction, int uiParam, int [] pvParam, int fWinIni);
+/** @method flags=dynamic */
 public static final native boolean SystemParametersInfoForDpi (int uiAction, int uiParam, NONCLIENTMETRICS pvParam, int fWinIni, int dpi);
 /**
  * @param lpKeyState cast=(PBYTE)

